src/humanloop/eval_utils/run.py

```diff
@@ -0,0 +1,751 @@
+    if "datapoints" not in dataset:
+        dataset["datapoints"] = []
+        # Use `upsert` to get existing dataset ID if no datapoints provided, given we can't `get` on path.
+        dataset["action"] = "add"
+    hl_dataset = client.datasets.upsert(
+        **dataset,
+    )
+    hl_dataset = client.datasets.get(
+        id=hl_dataset.id,
+        version_id=hl_dataset.version_id,
+        include_datapoints=True,
+    )
+
+    # Upsert the local Evaluators; other Evaluators are just referenced by `path` or `id`
+    local_evaluators: List[Evaluator] = []
+    if evaluators:
+        for evaluator in evaluators:
+            # If a callable is provided for an Evaluator, we treat it as External
+            eval_function = evaluator.get("callable")
+            if eval_function is not None:
+                # TODO: support the case where `file` logs generated on Humanloop but Evaluator logs generated locally
+                if function_ is None:
+                    raise ValueError(
+                        "Local Evaluators are only supported when generating Logs locally using your "
+                        f"{type_}'s `callable`. Please provide a `callable` for your file in order "
+                        "to run Evaluators locally."
+                    )
+                local_evaluators.append(evaluator)
+                spec = ExternalEvaluator(
+                    arguments_type=evaluator["args_type"],
+                    return_type=evaluator["return_type"],
+                    attributes={"code": inspect.getsource(eval_function)},
+                    evaluator_type="external",
+                )
+                client.evaluators.upsert(
+                    id=evaluator.get("id"),
+                    path=evaluator.get("path"),
+                    spec=spec,
+                )
+    function_ = typing.cast(Callable, function_)
+
+    # Validate upfront that the local Evaluators and Dataset fit
+    requires_target = False
+    for local_evaluator in local_evaluators:
+        if local_evaluator["args_type"] == "target_required":
+            requires_target = True
+            break
+    if requires_target:
+        missing_target = 0
+        for datapoint in hl_dataset.datapoints:  # type: ignore
+            if not datapoint.target:
+                missing_target += 1
+        if missing_target > 0:
+            raise ValueError(
+                f"{missing_target} Datapoints have no target. A target "
+                f"is required for the Evaluator: {local_evaluator['path']}"
+            )
+
+    # Get or create the Evaluation based on the name
+    evaluation = None
+    try:
+        evaluation = client.evaluations.create(
+            name=name,
+            evaluators=[{"path": e["path"]} for e in evaluators],  # type: ignore
+            file={"id": hl_file.id},
+        )
+    except ApiError as error_:
+        # If the name exists, go and get it # TODO: Update API GET to allow querying by name and file.
```

Comment:
should we block on adding this endpoint?

Reply:
this is already the way it works